File: samples/ios-demo/SceneViewDemo/Services/SketchfabAssetResolver.swift
import Foundation

/// Single-entry resolver that hands a sample/demo a ready-to-use file URL for a
/// `SketchfabSlug` — regardless of whether the model is cached, must be streamed,
/// or has to fall back to the bundled IPA asset.
///
/// Stage 1 deliverable for [#1152](https://github.com/sceneview/sceneview/issues/1152).
///
/// **Behavior (matches Android `SketchfabAssetResolver.kt` 1-for-1)**:
///  1. **Cache hit** → touch modification date (LRU bump) and return cached URL.
///  2. **Cache miss + API key present** → `SketchfabService.downloadModel`,
///     optionally validate via `boundsValidator`, evict LRU if cache exceeds
///     the budget, return the freshly-downloaded file URL.
///  3. **Cache miss + no key** → copy `SketchfabSlug.fallbackBundledPath` from
///     the main bundle into the cache directory under a `fallback-<uid>.usdz`
///     name and return that URL.
///  4. **Download fails OR validation rejects** → log the failure and fall back
///     to the bundled path as in case 3.
///
/// **Bounds validation.** The actor doesn't depend on RealityKit — the
/// `boundsValidator` parameter is `(URL, SketchfabSlug) async throws -> Bool`
/// and defaults to "accept-everything". Production callers wrap the resolver
/// with a validator backed by `Entity.visualBounds(recursive:relativeTo:)`.
/// This lets unit tests inject a fake validator and exercise the cache /
/// fallback paths without loading a real RealityKit scene.
///
/// **Threading.** The class is an `actor` so concurrent `resolve` calls
/// serialize through the same cache-mutation state. The underlying
/// `SketchfabService` is also an `actor`.
actor SketchfabAssetResolver {

    /// Stage-1 LRU cap — see Stage-1 spec on #1152.
    static let defaultCacheBudgetBytes: Int64 = 250 * 1024 * 1024

    /// Sentinel validator that accepts every download. Used in tests and as
    /// the default when no RealityKit validator is supplied.
    static let acceptAll: @Sendable (URL, SketchfabSlug) async throws -> Bool = { _, _ in true }

    private let cacheDirectoryName: String
    private let downloader: @Sendable (String) async throws -> URL
    private let bundleProvider: @Sendable (String) -> URL?
    private let boundsValidator: @Sendable (URL, SketchfabSlug) async throws -> Bool
    private let cacheBudgetBytes: Int64
    private let apiKeyProvider: @Sendable () -> String?

    /// Designated initializer. Tests pass custom closures; production callers
    /// stick with the defaults.
    ///
    /// - Parameters:
    ///   - downloader: Suspending download function `(uid) -> URL`. Defaults
    ///     to `SketchfabService.shared.downloadModel`.
    ///   - bundleProvider: Resolves a bundled fallback name to a `URL`.
    ///     Defaults to `Bundle.main.url(forResource:withExtension:)` matching.
    ///   - boundsValidator: Optional gate run after a successful download.
    ///   - cacheBudgetBytes: LRU cap enforced after each successful download.
    ///   - apiKeyProvider: Override for `SketchfabConfig.apiKey` (tests inject
    ///     `nil` to exercise the offline path without touching the environment).
    init(
        cacheDirectoryName: String = SketchfabConfig.cacheDirectoryName,
        downloader: @escaping @Sendable (String) async throws -> URL = { uid in
            try await SketchfabService.shared.downloadModel(uid: uid)
        },
        bundleProvider: @escaping @Sendable (String) -> URL? = { name in
            // Strip optional extension so callers can pass either
            // `animated_butterfly.usdz` or `animated_butterfly`.
            let lastDot = name.lastIndex(of: ".")
            let stem = lastDot.map { String(name[..<$0]) } ?? name
            let ext = lastDot.map { String(name[name.index(after: $0)...]) } ?? "usdz"
            return Bundle.main.url(forResource: stem, withExtension: ext)
        },
        boundsValidator: @escaping @Sendable (URL, SketchfabSlug) async throws -> Bool = acceptAll,
        cacheBudgetBytes: Int64 = defaultCacheBudgetBytes,
        apiKeyProvider: @escaping @Sendable () -> String? = { SketchfabConfig.apiKey }
    ) {
        self.cacheDirectoryName = cacheDirectoryName
        self.downloader = downloader
        self.bundleProvider = bundleProvider
        self.boundsValidator = boundsValidator
        self.cacheBudgetBytes = cacheBudgetBytes
        self.apiKeyProvider = apiKeyProvider
    }

    /// Resolve `slug` to a local file URL ready for `Entity(contentsOf:)`.
    ///
    /// Never throws. Errors are logged + the bundled fallback is returned so
    /// the rendering pipeline always has something to consume.
    func resolve(_ slug: SketchfabSlug) async throws -> URL {
        // 1. Cache hit short-circuit.
        let cached = try cacheFileURL(for: slug.uid)
        if FileManager.default.fileExists(atPath: cached.path),
           (try? FileManager.default.attributesOfItem(atPath: cached.path)[.size] as? NSNumber)?.intValue ?? 0 > 0 {
            touch(cached)
            return cached
        }

        // 2. No key → bundled fallback.
        guard let apiKey = apiKeyProvider(), !apiKey.isEmpty else {
            return try extractBundledFallback(slug)
        }
        _ = apiKey

        // 3. Try the network path.
        let downloaded: URL
        do {
            downloaded = try await downloader(slug.uid)
        } catch {
            print("[SketchfabResolver] download failed for \(slug.uid) (\(error)); using bundled fallback")
            return try extractBundledFallback(slug)
        }

        // 4. Bounds-check + animation-flag check.
        let passes: Bool
        do {
            passes = try await boundsValidator(downloaded, slug)
        } catch {
            print("[SketchfabResolver] bounds validator threw for \(slug.uid): \(error); treating as reject")
            try? FileManager.default.removeItem(at: downloaded)
            return try extractBundledFallback(slug)
        }
        if !passes {
            print("[SketchfabResolver] bounds validation rejected \(slug.uid); falling back")
            try? FileManager.default.removeItem(at: downloaded)
            return try extractBundledFallback(slug)
        }

        // 5. Enforce LRU budget and return.
        evictLruIfOverBudget(maxBytes: cacheBudgetBytes)
        return downloaded
    }

    /// Walk the Sketchfab cache directory, sort by modification date ascending,
    /// and delete oldest entries until total size ≤ `maxBytes`. Used both
    /// internally after every successful download and exposed as a test /
    /// maintenance hook.
    func evictLruIfOverBudget(maxBytes: Int64? = nil) {
        let target = maxBytes ?? cacheBudgetBytes
        guard let root = try? cacheRoot() else { return }
        let fm = FileManager.default
        guard let entries = try? fm.contentsOfDirectory(
            at: root,
            includingPropertiesForKeys: [.fileSizeKey, .contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ) else { return }
        let files: [(URL, Int64, Date)] = entries.compactMap { url in
            let values = try? url.resourceValues(forKeys: [.fileSizeKey, .contentModificationDateKey])
            let size = Int64(values?.fileSize ?? 0)
            let mtime = values?.contentModificationDate ?? .distantPast
            return (url, size, mtime)
        }
        var total = files.reduce(Int64(0)) { $0 + $1.1 }
        guard total > target else { return }
        let sorted = files.sorted { $0.2 < $1.2 }
        for (url, size, _) in sorted {
            if total <= target { break }
            if (try? fm.removeItem(at: url)) != nil {
                total -= size
            }
        }
    }

    // MARK: - Internals

    func cacheRoot() throws -> URL {
        let caches = try FileManager.default.url(
            for: .cachesDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let dir = caches.appendingPathComponent(cacheDirectoryName, isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    func cacheFileURL(for uid: String) throws -> URL {
        try cacheRoot().appendingPathComponent("\(uid).glb")
    }

    /// Copy `slug.fallbackBundledPath` from the main bundle into the cache
    /// directory and return the resulting URL. The destination is named
    /// `fallback-<uid>.usdz` so it can't accidentally pose as a real Sketchfab
    /// download (LRU bumps target real downloads only).
    func extractBundledFallback(_ slug: SketchfabSlug) throws -> URL {
        let root = try cacheRoot()
        let dest = root.appendingPathComponent("fallback-\(slug.uid).usdz")
        if FileManager.default.fileExists(atPath: dest.path) {
            if let size = try? FileManager.default.attributesOfItem(atPath: dest.path)[.size] as? NSNumber,
               size.intValue > 0 {
                return dest
            }
        }
        guard let source = bundleProvider(slug.fallbackBundledPath) else {
            throw SketchfabResolverError.missingBundledFallback(
                slug: slug.uid,
                path: slug.fallbackBundledPath
            )
        }
        try? FileManager.default.removeItem(at: dest)
        try FileManager.default.copyItem(at: source, to: dest)
        return dest
    }

    private nonisolated func touch(_ url: URL) {
        try? FileManager.default.setAttributes(
            [.modificationDate: Date()],
            ofItemAtPath: url.path
        )
    }
}

/// Errors surfaced by `SketchfabAssetResolver` — these only escape when the
/// resolver can't even produce a fallback (e.g. the bundled asset was
/// stripped at build time), which is a build-config bug.
enum SketchfabResolverError: Error, LocalizedError {
    case missingBundledFallback(slug: String, path: String)

    var errorDescription: String? {
        switch self {
        case .missingBundledFallback(let slug, let path):
            return "Missing bundled fallback at Bundle.main/\(path) for slug \(slug). "
                + "Add the file to the Xcode target or fix the SketchfabSlug entry."
        }
    }
}

File: samples/android-demo/src/main/java/io/github/sceneview/demo/sketchfab/SketchfabAssetResolver.kt
package io.github.sceneview.demo.sketchfab

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException

/**
 * Single-entry resolver that hands a sample/demo a ready-to-use `File` for a
 * [SketchfabSlug] — regardless of whether the model is cached, must be streamed,
 * or has to fall back to the bundled APK asset.
 *
 * Stage 1 deliverable for [#1152](https://github.com/sceneview/sceneview/issues/1152).
 *
 * **Behavior (matches iOS `SketchfabAssetResolver.swift` 1-for-1)**:
 *  1. **Cache hit** → touch `lastModified` (LRU bump) and return the cached file.
 *  2. **Cache miss + API key present** → [SketchfabService.downloadModel],
 *     optionally validate via [boundsValidator], evict LRU if cache exceeds the
 *     budget, return the freshly-downloaded file.
 *  3. **Cache miss + no key** → copy [SketchfabSlug.fallbackBundledPath] from
 *     `assets/` into the cache directory under a `fallback-<uid>.glb` name and
 *     return that file. The caller still gets a real `File` so the demo can
 *     render — Sketchfab is just an invisible CDN.
 *  4. **Download fails OR validation rejects** → log the failure (telemetry
 *     hook later) and fall back to the bundled path as in case 3.
 *
 * **Bounds validation.** The class doesn't depend on Filament — the
 * [boundsValidator] parameter is a function of `(file, slug) → Boolean` and
 * defaults to "accept-everything". The production callers wrap the resolver
 * with a validator backed by `ModelLoader` (which must be on the main thread,
 * see [io.github.sceneview.loaders.ModelLoader]). This lets unit tests inject
 * a fake validator and exercise the cache/fallback paths without a Filament
 * `Engine`.
 *
 * **Threading.** All work is dispatched to [Dispatchers.IO]. The resolver is
 * stateless past its constructor — multiple instances share the same on-disk
 * cache because [SketchfabService] is a process-wide singleton.
 *
 * @property context             The application context; only used for
 *                               `cacheDir` and `assets` access.
 * @property downloader          Suspending download function `(uid) -> File`.
 *                               Defaults to [SketchfabService.downloadModel]
 *                               via the process singleton. Tests substitute a
 *                               stub that writes a synthetic file into the
 *                               cache directory without hitting the network.
 * @property boundsValidator     Optional gate run after a successful download.
 *                               Return `false` to discard the file and fall
 *                               back. Default: accept-everything.
 * @property cacheBudgetBytes    LRU cap enforced after each successful
 *                               download. Defaults to 250 MB per the Stage-1
 *                               spec — lower than [SketchfabConfig.CACHE_MAX_BYTES]
 *                               so the demo APK never fills the user's cache
 *                               partition with model data.
 */
class SketchfabAssetResolver(
    private val context: Context,
    /**
     * Coroutine that fetches a model by uid and returns the cached `File`.
     *
     * Default delegates to the singleton [SketchfabService.downloadModel].
     * Tests substitute a stub that writes a synthetic file into the cache
     * directory without hitting the network.
     */
    private val downloader: suspend (String) -> File = { uid ->
        SketchfabService.getInstance(context).downloadModel(uid)
    },
    private val boundsValidator: suspend (File, SketchfabSlug) -> Boolean = ACCEPT_ALL,
    private val cacheBudgetBytes: Long = DEFAULT_CACHE_BUDGET_BYTES,
) {
    /**
     * Resolve [slug] to a local `File` suitable for `rememberModelInstance(modelLoader, file.absolutePath)`.
     *
     * Never throws. Errors are logged + the bundled fallback is returned so
     * the rendering pipeline always has something to consume.
     */
    suspend fun resolve(slug: SketchfabSlug): File = withContext(Dispatchers.IO) {
        // 1. Cache hit short-circuit. SketchfabService.downloadModel() already
        //    bumps lastModified, but we check here first to avoid even building
        //    the request when the key is absent.
        val cached = cacheFile(slug.uid)
        if (cached.exists() && cached.length() > 0L) {
            cached.setLastModified(System.currentTimeMillis())
            return@withContext cached
        }

        // 2. No key → straight to bundled fallback. No telemetry — this is
        //    the documented offline / first-launch path.
        val apiKey = SketchfabConfig.apiKey
        if (apiKey.isNullOrBlank()) {
            return@withContext extractBundledFallback(slug)
        }

        // 3. Try the network path. Any failure logs + falls back. We never let
        //    a network error bubble into the demo composable.
        val downloaded = try {
            downloader(slug.uid)
        } catch (e: SketchfabService.SketchfabError) {
            Log.w(TAG, "Sketchfab download failed for ${slug.uid} (${e.message}); using bundled fallback")
            return@withContext extractBundledFallback(slug)
        } catch (e: IOException) {
            Log.w(TAG, "IO error downloading ${slug.uid}; using bundled fallback", e)
            return@withContext extractBundledFallback(slug)
        }

        // 4. Bounds-check + animation-flag check (callers may have wired up the
        //    Filament-backed validator). If the model fails, delete the bad
        //    download so a retry doesn't re-use the corrupt blob, then fall back.
        val passes = try {
            boundsValidator(downloaded, slug)
        } catch (t: Throwable) {
            Log.w(TAG, "Bounds validator threw for ${slug.uid}; treating as reject", t)
            false
        }
        if (!passes) {
            Log.w(TAG, "Bounds validation rejected ${slug.uid}; falling back to bundled asset")
            runCatching { downloaded.delete() }
            return@withContext extractBundledFallback(slug)
        }

        // 5. Successful path: enforce the LRU budget and return.
        evictLruIfOverBudget(cacheBudgetBytes)
        downloaded
    }

    /**
     * Walk the Sketchfab cache directory, sort by `lastModified` ascending, and
     * delete oldest entries until total size ≤ [maxBytes]. Used both internally
     * after every successful download and exposed as a test/maintenance hook.
     *
     * No-op when the directory doesn't exist yet or is already under budget.
     */
    fun evictLruIfOverBudget(maxBytes: Long = cacheBudgetBytes) {
        val root = File(context.cacheDir, SketchfabConfig.CACHE_DIR_NAME)
        if (!root.exists()) return
        val files = root.listFiles()?.filter { it.isFile } ?: return
        var total = files.sumOf { it.length() }
        if (total <= maxBytes) return
        val sorted = files.sortedBy { it.lastModified() }
        for (file in sorted) {
            if (total <= maxBytes) break
            val size = file.length()
            if (file.delete()) total -= size
        }
    }

    // ── Internals ─────────────────────────────────────────────────────────

    /**
     * Where [SketchfabService] caches downloaded GLBs. Kept identical to the
     * service's own layout (`<cacheDir>/sketchfab/<uid>.glb`) so a download done
     * by another code path is immediately visible here as a "cache hit".
     */
    internal fun cacheFile(uid: String): File =
        File(File(context.cacheDir, SketchfabConfig.CACHE_DIR_NAME), "$uid.glb")

    /**
     * Copy [SketchfabSlug.fallbackBundledPath] from `assets/` into the cache
     * directory and return the resulting file. The destination is named
     * `fallback-<uid>.glb` so it can't accidentally pose as a real Sketchfab
     * download (LRU bumps target real downloads only).
     *
     * Throws [IllegalStateException] if the bundled asset is missing — that's
     * a build-time bug, not a runtime degradation we can absorb.
     */
    @Throws(IllegalStateException::class)
    internal fun extractBundledFallback(slug: SketchfabSlug): File {
        val root = File(context.cacheDir, SketchfabConfig.CACHE_DIR_NAME).apply { mkdirs() }
        val dest = File(root, "fallback-${slug.uid}.glb")
        if (dest.exists() && dest.length() > 0L) return dest
        try {
            context.assets.open(slug.fallbackBundledPath).use { input ->
                dest.outputStream().use { output -> input.copyTo(output) }
            }
        } catch (e: IOException) {
            throw IllegalStateException(
                "Missing bundled fallback at assets/${slug.fallbackBundledPath} for slug ${slug.uid}",
                e,
            )
        }
        return dest
    }

    companion object {
        private const val TAG = "SketchfabResolver"

        /** Stage-1 LRU cap — see Stage-1 spec on #1152. */
        const val DEFAULT_CACHE_BUDGET_BYTES: Long = 250L * 1024 * 1024

        /** Sentinel validator that accepts every download. Used in tests and as
         *  the default when no Filament validator is supplied. */
        val ACCEPT_ALL: suspend (File, SketchfabSlug) -> Boolean = { _, _ -> true }

        /**
         * Reference implementation of a bounds validator for production callers
         * that already hold a `ModelLoader`. Lives in a separate file so this
         * one stays free of Filament dependencies — see
         * `SketchfabBoundsValidator.kt` (Stage 2).
         */
    }
}
